src/stdlib_sorting_sort_index.fypp

```diff
@@ -0,0 +1,873 @@
+!!   http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+!!   <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+!!   option. This file may not be copied, modified, or distributed
+!!   except according to those terms.
+!!
+!! so the license for the original`slice.rs` code is compatible with the use
+!! of modified versions of the code in the Fortran Standard Library under
+!! the MIT license.
+
+submodule(stdlib_sorting) stdlib_sorting_sort_index
+
+    implicit none
+
+contains
+
+#:for k1, t1 in IRS_KINDS_TYPES
+
+    module subroutine ${k1}$_sort_index( array, index, work, iwork, reverse )
+! A modification of `${k1}$_ord_sort` to return an array of indices that
+! would perform a stable sort of the `ARRAY` as input, and also sort `ARRAY`
+! as desired. The indices by default
+! correspond to a non-decreasing sort, but if the optional argument
+! `REVERSE` is present with a value of `.TRUE.` the indices correspond to
+! a non-increasing sort. The logic of the determination of indexing largely
+! follows the `"Rust" sort` found in `slice.rs`:
+! https://github.com/rust-lang/rust/blob/90eb44a5897c39e3dff9c7e48e3973671dcd9496/src/liballoc/slice.rs#L2159
+! The Rust version is a simplification of the Timsort algorithm described
+! in https://svn.python.org/projects/python/trunk/Objects/listsort.txt, as
+! it drops both the use of 'galloping' to identify bounds of regions to be
+! sorted and the estimation of the optimal `run size`. However it remains
+! a hybrid sorting algorithm combining an iterative Merge sort controlled
+! by a stack of `RUNS` identified by regions of uniformly decreasing or
+! non-decreasing sequences that may be expanded to a minimum run size, with
+! an insertion sort.
+!
+! Note the Fortran implementation simplifies the logic as it only has to
+! deal with Fortran arrays of intrinsic types and not the full generality
+! of Rust's arrays and lists for arbitrary types. It also adds the
+! estimation of the optimal `run size` as suggested in Tim Peters'
+! original listsort.txt, and the optional `work` and `iwork` arraya to be
+! used as scratch memory.
+
+        ${t1}$, intent(inout)                    :: array(0:)
+        integer(int_size), intent(out)           :: index(0:)
+        ${t1}$, intent(out), optional            :: work(0:)
+        integer(int_size), intent(out), optional :: iwork(0:)
+        logical, intent(in), optional            :: reverse
+
+        integer(int_size) :: array_size, i, stat
+        ${t1}$, allocatable :: buf(:)
+        integer(int_size), allocatable :: ibuf(:)
+
+        array_size = size(array, kind=int_size)
+
+        do i = 0, array_size-1
+            index(i) = i+1
+        end do
+
+        if ( present(reverse) ) then
+            if ( reverse ) then
+                call reverse_segment( array, index )
+            end if
+        end if
+
+! If necessary allocate buffers to serve as scratch memory.
+        if ( present(work) ) then
```

Comment:
Should there be a check on the size of `work`?

Reply:
Yes.